HashTable/hashtable.c

```diff
@@ -0,0 +1,193 @@
+    }
+    table->size = (unsigned int)size;
+    table->elements = (Node**)malloc(table->size*sizeof(Node*));
+
+    if (table->elements == NULL) {
+        perror("Memory allocate error");
+        free(table);
+        return NULL;
+    }
+
+    for (int i = 0; i < table->size; ++i) {
+        table->elements[i] = NULL;
+    }
+
+    return table;
+}
+
+unsigned int hashFunction(char* word, int length, unsigned int tableSize) {
+    unsigned int hash = 0;
+    for (int i = 0; i < length ; i++) {
+        hash += (unsigned int)word[i];
+    }
+    return hash % tableSize;
+}
+
+HashTable* getNewHashTable(HashTable* table, int previousSize) {
+    unsigned int newSize = previousSize * 2;
+    HashTable* newTable = initTable(newSize);
+
+    if (newTable == NULL) {
+        return NULL;
+    }
+
+    for (int i = 0; i < previousSize; i++) {
+        Node* node = (table->elements)[i];
+
+        if (node != NULL) {        
+            char* word = getNodeWord(node);
+            unsigned int newHash = hashFunction(word, strlen(word), newSize);
+            newTable->elements[newHash] = node;
+        }
+    }
+    free(table->elements);
+    free(table);
+    return newTable;
+}
+
+bool addWord(HashTable** tablePointer, char* word) {
+    HashTable* table = *tablePointer;
+    int length = strlen(word);
+    unsigned int index = hashFunction(word, length, getTableSize(table));
+    table->elements[index] = insertOrIncreaseCounter((table->elements)[index], word);
+    double loadFactor = calculateLoadFactor(table);
+
+    if (loadFactor > 0.7) {
+        HashTable* newTable = getNewHashTable(table, table->size);
+
+        if (newTable == NULL) {
+            return false;
+        }
+
+        *tablePointer = newTable;
+    }
+    return true;
+}
+
+void printTable(HashTable* table) {
+    for (int i = 0; i < table->size; i++) {
+        printList((table->elements)[i]);
+    }
+}
+
+void freeTable(HashTable** table) {
+    for (int i = 0; i < (*table)->size; i++) {
+        if (((*table)->elements)[i] != NULL) {
+            freeList(&((*table)->elements)[i]);
+        }
+    }
+    free((*table)->elements);
+    free(*table);
+    *table = NULL;
+}
+
+double calculateLoadFactor(HashTable* table) {
+    int nonEmptySegments = 0;
+    for (int i = 0; i < table->size; i++) {
+        if ((table->elements)[i] != NULL) {
+            nonEmptySegments++;
+        }
+    }
+    return (double)nonEmptySegments / table->size;
+}
+
+double calculateAvgListLength(HashTable* table) {
+    int totalLength = 0;
+    int nonEmptySegments = 0;
+    for (int i = 0; i < table->size; i++) {
+        int length = 0;
+        Node* current = (table->elements)[i];
+        while (current) {
+            length++;
+            current = getNext(current);
+        }
+        if (length > 0) {
+            totalLength += length;
+            nonEmptySegments++;
+        }
+    }
+    return nonEmptySegments ? (double)totalLength / nonEmptySegments : 0.0;
```

Comment:
```suggestion
    return nonEmptySegments != 0 ? (double)totalLength / nonEmptySegments : 0.0;
```
Неявных преобразований типов по возможности стоит избегать, они источник ошибок